ThingWorx.agent.lib.nut

```diff
@@ -0,0 +1,390 @@
+    function _processResponse(response, callback) {
+        _logDebug(format("Response status: %d, body: %s", response.statuscode, response.body));
+        local errType = null;
+        local errDetails = null;
+        local httpStatus = response.statuscode;
+        if (httpStatus < 200 || httpStatus >= 300) {
+            errType = THING_WORX_ERROR.REQUEST_FAILED;
+            errDetails = format("%s: %i", THING_WORX_REQUEST_FAILED, httpStatus);
+            try {
+                response.body = (response.body == "") ? {} : http.jsondecode(response.body);
+            } catch (e) {
+                _logError(e);
+            }
+        }
+
+        local error = errType ? ThingWorxError(errType, errDetails, httpStatus, response.body) : null;
+        _invokeDefaultCallback(callback, error);
+    }
+
+    // Validates the argument is not empty. Invokes callback with THING_WORX_ERROR.LIBRARY_ERROR if the check failed.
+    function _validateNonEmptyArg(arg, argName, callback, invokeCallback = null) {
+        if (arg == null || typeof arg == "string" && arg.len() == 0) {
+            local error = ThingWorxError(
+                THING_WORX_ERROR.LIBRARY_ERROR,
+                format("%s: %s", THING_WORX_NON_EMPTY_ARG, argName));
+            if (!invokeCallback) {
+                invokeCallback = _invokeDefaultCallback;
+            }
+            invokeCallback(callback, error);
+            return false;
+        }
+        return true;
+    }
+
+    // Validates the arguments are not empty. Invokes callback with THING_WORX_ERROR.LIBRARY_ERROR if the check failed.
+    function _validateNonEmptyArgs(args, callback, invokeCallback = null) {
+        foreach (argName, arg in args) {
+            if (!_validateNonEmptyArg(arg, argName, callback, invokeCallback)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    // Logs error occurred and invokes default callback with single error parameter.
+    function _invokeDefaultCallback(callback, error) {
+        if (error) {
+            _logError(error.details);
+        }
+        if (callback) {
+            imp.wakeup(0, function () {
+                callback(error);
+            });
+        }
+    }
+
+    // Invokes existThing method callback.
+    function _invokeExistCallback(callback, error, exist = false) {
+        if (callback) {
+            imp.wakeup(0, function () {
+                callback(error, exist);
+            });
+        }
+    }
+
+    // Logs error occurred during the library methods execution.
+    function _logError(message) {
+        if (_debug) {
+            server.error("[ThingWorx] " + message);
+        }
+    }
+
+    // Logs debug messages occurred during the library methods execution.
+    function _logDebug(message) {
+        if (_debug) {
+            server.log("[ThingWorx] " + message);
+        }
+    }
+}
+
+// Auxiliary class, represents error returned by the library.
+class ThingWorxError {
```

Comment:
I would define this class inside ThingWorx, i.e.

```
class ThingWorx {
    Error = class {}
}
```

So that you can access it `ThingWorx.Error`

Reply:
Ok, done